overlord/snapstate/handlers.go

```diff
@@ -1790,6 +1800,10 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	if oldCandidateIndex < 0 {
 		snapst.Sequence = append(snapst.Sequence[:currentIndex], snapst.Sequence[currentIndex+1:]...)
 	} else if !isRevert {
+		// account for revisions discarded before the install failed
+		discarded := countExcludedRevs(oldRevsBeforeCand, snapst.Sequence)
+		oldCandidateIndex -= discarded
```

Comment:
I think if this runs an old task, then `discarded` is zero and the code is essentially what it was before (so it won't break where it didn't before). I'm not sure we can avoid the previous panic cases without the extra info but I think having an old task with a new snapd should be rare since there would have to be an update between doLink and undoLink